### ISSUE TITLE
Error in accounts with high instance churn / SSM Inventory latency

### ISSUE THREAD

**Issue by jantman** [mludvig/aws-ssm-tools#7]
There appears to be some latency between when an instance is terminated and when it's removed from the SSM inventory. As a result, in accounts with a large number of instances and high churn (I have ~1,000 instances in a single region, and have hundreds of them are replaced every day), it's likely that DescribeInstances will fail on one or more instances that are still in the SSM Inventory. In these cases, aws-ssm-tools' method of calling DescribeInstances with the full list of instance IDs in the SSM Inventory will render it useless as it will encounter an error.

The following is a trimmed-down excerpt of the ``ssm-session`` debug output from an account that has 724 instances in the SSM Inventory for this region. 649 of those were properly recognized by SSM Inventory as terminated and skipped (``DEBUG: Ignoring terminated instance:``), but it looks like 4 of them were terminated within a short enough period that SSM doesn't know they're gone, but DescribeInstances errors:

```
[ssm-session] DEBUG: Logging level set to DEBUG
[ssm-session] DEBUG: Fetching SSM inventory
...
[ssm-session] DEBUG: Added instance: i-02df27b30a98a2bf6: {'InstanceId': 'i-02df27b30a98a2bf6', 'HostName': 'ip-10-230-8-231'}
...
[ssm-session] DEBUG: Added instance: i-06a5cd08dccd6cf94: {'InstanceId': 'i-06a5cd08dccd6cf94', 'HostName': 'ip-10-230-8-70'}
...
[ssm-session] DEBUG: Added instance: i-093177e4964fd4be3: {'InstanceId': 'i-093177e4964fd4be3', 'HostName': 'ip-10-142-0-108'}
...
[ssm-session] DEBUG: Added instance: i-0f4d372b4cba55a1a: {'InstanceId': 'i-0f4d372b4cba55a1a', 'HostName': 'ip-10-230-12-196'}
...
[ssm-session] ERROR: An error occurred (InvalidInstanceID.NotFound) when calling the DescribeInstances operation: The instance IDs 'i-02df27b30a98a2bf6, i-06a5cd08dccd6cf94, i-093177e4964fd4be3, i-0f4d372b4cba55a1a' do not exist
```

I can think of a few possible fixes for this:

1. Right now, if the argument to ssm-session is an Instance ID (``i-`` string), we short-circuit InstanceResolver.resolve_instance() and directly return that Instance ID. The SSM Inventory data includes both Instance IDs and HostNames, but we make the full DescribeInstances call any time our argument isn't an Instance ID. In InstanceResolver.get_list() we could check for hostname matches, and if there are any, return those without needing to call DescribeInstances to get the IP addresses. This would both fix this specific bug in all cases except specifying an IP address, and would also avoid potentially-slow DescribeInstances calls in cases where we don't need them.
2. We could switch from calling DescribeInstances with all of the SSM Inventory instance IDs to using the boto3 ``EC2.Instance(instance_id)`` resource for each instance. This would greatly increase the number of DescribeInstances API calls, and likely adversely effect performance, but would allow us to try/except on each instance ID itself.
3. Since boto3 captures detailed information on the error, and AWS error messages are _usually_ fixed and reliable strings, we could catch the ``InvalidInstanceID.NotFound`` error, parse out the Instance IDs from the error message, remove them from the list of instances, and retry the DescribeInstances operation with the updated list of instance IDs.

I'd be happy to open a pull request to implement any of these options, or a combination of them, but I'll leave the choice of which up to you. Since this should probably be an edge case, I'd personally lean towards a combination of options 1 and 3.

(BTW - This is an **absolutely wonderful** project! Thank you so much! It's completely revolutionized how I work with EC2 instances when I need access to an actual instance, and I'm working hard to let everyone else at my company know about it.)

**Comment by jantman** [comment]
FYI I've opened PR #10 which has a fix for this that seems to be working for me.

**Comment by mludvig** [comment]
Hi, as explained in your PR it's now fixed in #16, please test.
Thanks for your contribution!

**Comment by jantman** [comment]
#16 confirmed as fix, thanks so much!!!
